streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java

```diff
@@ -410,18 +410,6 @@ public final void addSource(final Topology.AutoOffsetReset offsetReset,
             }
         }
 
-        for (final Pattern otherPattern : earliestResetPatterns) {
-            if (topicPattern.pattern().contains(otherPattern.pattern()) || otherPattern.pattern().contains(topicPattern.pattern())) {
-                throw new TopologyException("Pattern " + topicPattern + " will overlap with another pattern " + otherPattern + " already been registered by another source");
-            }
-        }
-
-        for (final Pattern otherPattern : latestResetPatterns) {
-            if (topicPattern.pattern().contains(otherPattern.pattern()) || otherPattern.pattern().contains(topicPattern.pattern())) {
-                throw new TopologyException("Pattern " + topicPattern + " will overlap with another pattern " + otherPattern + " already been registered by another source");
-            }
-        }
-
```

Comment:
Saw this and at first I thought it was broken because it only considers pattern-subscribed topics that happened to explicitly configure an offset reset policy. Unless I'm missing something here, that makes no sense and we should consider _all_  source patterns and whether they overlap.
But then I started thinking, why does it matter if they overlap? Just because one pattern is a substring of another does not mean that they'll match the same topics. So I think that we should actually just remove this restriction altogether. Am I missing anything here?

Reply:
I agree on the first part.
Regarding the second part, I had similiar thoughts when I wrote my comment in `mergeDuplicateSourceNodes()`.
But I might also be missing something here.

Reply:
I might _also_ be missing something, but what's the scenario where one pattern is a substring of another and they _dont_ match the same topics? If you take Bruno's example from earlier of `topic*` and `topi*`, `topi*` would be considered a substring of `topic*` and they would both match `topic A`, right? I guess the other scenario is if we have a topic `topia A`, that would match `topi*` and not `topic*`. So I guess it seems like it isn't always true that they'll overlap, but we would want to check if they do, right?

Reply:
Pattern `topic*` is contained in pattern `topic*A`. However, `topic*A` matches only a subset of `topic*`. So, they do not match exactly the same topics. But matching exactly the same topics is a pre-requisite for merging the source nodes.

Reply:
I think in this case we were matching whether the pattern's string was a literal substring of another pattern's string, not whether the regexes themselves are substrings. So `topi*` would not be a substring of `topic*` because `topi*` is not contained literally within the string `topic*`. It's not doing a smart regex-matching, just a dumb  literal string comparison